src/time_tracker/invoices/core.clj

```diff
@@ -1,49 +1,149 @@
+(defn invoice
+  "Constructs an invoice"
+  [{:keys [users timers user-id->rate] :as invoice-params}]
+  {:pre  [(seq users)]
+   :post [(s/valid? ::invoices-spec/invoice %)]}
+  (merge (select-keys invoice-params [:start :end :client :address
+                                      :notes :tax-rates :currency :utc-offset])
+         {:user-hours (user-hours users timers user-id->rate)}))
 
-(defn generate-csv
-  "Generates a CSV given `users` and `timers`.
-  There will be a row for every user in `users`."
-  [users timers]
-  (let [user-id->name  (id->name users)
-        user-id->hours (build-user-id->hours (keys users) timers)]
-    (with-out-str
-      (csv/write-csv *out*
-                     (->> (csv-rows user-id->hours user-id->name)
-                          (cons ["Name" "Hours Logged"]))))))
+;; TODO: Map these to their symbols and render ₹ properly
+(def currency-symbols
+  {:usd "USD "
+   :inr "INR "})
+
+(defn- money-str
+  [currency amount]
+  (str (currency-symbols currency) amount))
+
+(defn- invoice-items
+  [{:keys [user-hours currency]} user-id->name]
+  (for [{:keys [id hours rate]} user-hours]
+    {:name   (user-id->name id)
+     :hours  hours
+     :rate   rate
+     :amount (util/round-to-two-places (* hours rate))}))
+
+(defn- subtotal
+  [invoice-items]
+  {:pre  [(s/valid? ::invoices-spec/items invoice-items)]
+   :post [(s/valid? ::core-spec/money-val %)]}
+  (->> invoice-items
+       (map :amount)
+       (apply +)))
+
+(defn- tax-amounts
+  [taxes subtotal-amount]
+  (for [{:keys [tax-name tax-percentage]} taxes]
+    {:name       tax-name
+     :amount     (-> (* subtotal-amount (util/divide-money tax-percentage 100.00M))
+                     (util/round-to-two-places))
+     :percentage tax-percentage}))
+
+(defn- grand-total
+  [subtotal-amount tax-maps]
+  {:pre  [(s/valid? ::core-spec/money-val subtotal-amount)
+          (s/valid? ::invoices-spec/tax-amounts tax-maps)]
+   :post [(s/valid? ::core-spec/money-val %)]}
+  (-> (apply + subtotal-amount (map :amount tax-maps))))
+
+(defn printable-invoice
+  [{:keys [tax-rates start end utc-offset] :as invoice} user-id->name]
+  (let [items           (invoice-items invoice user-id->name)
+        subtotal-amount (subtotal items)
+        taxes           (tax-amounts tax-rates subtotal-amount)
+        amount-due      (grand-total subtotal-amount taxes)
+        from-date       start
+        ;; `end` is exclusive, but the to-date must be inclusive.
+        to-date         (-> (util/epoch->clj-time end utc-offset)
+                            (time/minus (time/days 1))
+                            (util/to-epoch-seconds))]
+    (def *gt-output amount-due)
+    (merge (select-keys invoice [:client :address :currency :utc-offset :notes])
+           {:items items
+            :subtotal subtotal-amount
+            :amount-due amount-due
+            :from-date from-date
+            :to-date to-date
+            :tax-amounts taxes})))
```

Comment:
It's a very minor thing, but I would inline the computation of independent values into the map's values. It reduces the overhead of looking back to the computation from the values a bit.